tests/MongoDB.Bson.Tests/Specifications/bson-corpus/BsonCorpusTestRunner.cs

```diff
@@ -0,0 +1,316 @@
+﻿/* Copyright 2020 MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+using FluentAssertions;
+using MongoDB.Bson.IO;
+using MongoDB.Bson.Serialization;
+using MongoDB.Bson.Serialization.Serializers;
+using MongoDB.Bson.TestHelpers.JsonDrivenTests;
+using Xunit;
+
+namespace MongoDB.Bson.Tests.Specifications.bson_corpus
+{
+    public class BsonCorpusTestRunner
+    {
+        [SkippableTheory]
+        [ClassData(typeof(TestCaseFactory))]
+        public void RunTestDefinition(JsonDrivenTestCase testCase)
+        {
+            var definition = testCase.Test;
```

Comment:
>It is not necessary here
I don't think so.

>Each of RunValidTest, RunDecodeErrorsTest, RunParseErrorsTest has it's own EnsureAllFieldsAreValid call with narrowed fields list encountered in respective section.

it's true, but all of these methods validate nested nodes, not the top one. In other words, you don't validate the top-level nodes:
![image](https://user-images.githubusercontent.com/44870738/80844632-cd995e80-8c0f-11ea-9f04-d2c40223e443.png)

Reply:
Also, I just noticed, it looks like you don't even process these values?

Reply:
Yes, top level values are actually just skipped. I don't see the specification ([link](https://github.com/mongodb/specifications/blob/master/source/bson-corpus/bson-corpus.rst#top-level-keys)) enforcing any check against top level fields. And in method `RunTestDefinition` top level keys are already absent as `testCase` contains only test from `valid`, `parseErrors` or `decodeErrors` sections.
@rstam considering the above, should we assert top level fields `bson_type` and `test_key`?

Reply:
Looks like the `bson_type` and `test_key` fields are mostly descriptive.

They are indirectly tested by ensuring that after parsing and write back out the valid values (to both BSON and JSON) the correct result is obtained.

Not sure there is anything worthwhile to assert against these fields.

Reply:
I'm not sure that the key point should we assert or not. It can be: should we use this in tests or no.

Reply:
Use how?

Reply:
Idk yet. I guess this is an explanation: `this will be the synthetic value "0x00" for "whole document" tests like top.json.`, but I don'really understand it.

Reply:
NOTE: in any case these values should be part of `JsonDrivenHelper.EnsureAllFieldsAreValid(..` assertation

Reply:
Also, maybe these lines can be helpful: https://github.com/mongodb/mongo-python-driver/commit/b6c74cf33dd5b38c8aa6d08b8a8e6b0073178cad#diff-b0089c4ea9a303b280442d9c6d18888cR73 (I just found it, but didn't look closely)

Reply:
I agree with Dmitry that we should validate the field names (even for fields like `bson_type` and `test_key` that we actually ignore).

Reply:
I guess also we can investigate how these fields are used in the python implementation.

Reply:
I implemented this by copying the fields in question into test cases and validating their presence in the test run.
Done.

Reply:
@rstam , can you please take a look at my above comment regarding the fact that some fields in json files (which we currently ignore) actually can be processed (at least in python). Not sure that it's 100% required, but I don't see a reason why we should ignore them (at least yet)

Reply:
I think `bson_type` and `test_key` are actually redundant (descriptive only) and can be safely ignored.

I also can't think of anything worthwhile to assert against using them.

If the actual test content was not in agreement with `bson_type` and `test_key` it would be a bad JSON file, not a problem with our test runner.

Reply:
Yep, as far as I found, previously `test_key` had a bit different description: 
>* ``test_key``: name of a field in a ``valid`` test case ``extjson`` document
  should be checked against the case's ``string`` field.

and I guess this is the reason why some drivers use this value in assertions. Now I confirmed with David that it's only for diagnostic reasons.

As for the `bson_type`. Looking at the python implementation, I guess they can be used to decide whether we need or don't need to make asserting (full asserting?) But It's probably not the case for us. I guess the same case is about `deprecated` value.

Reply:
So, resolved.